### PR TITLE
Set job to failed when task raises and job is aborting (regardless retry)

### LINKED ISSUE #1131: Retrying a task, that has been aborted
Hi guys!

I've been doing some manual QA for a project I'm working on and found some issues with the newly added `abort` logic for task cancellation. By the way, a neat feature I was waiting for, glad it has been added!

Here's the general context - I have a single type of task for the procrastinate, and I use a retry strategy over it to handle some expected exceptions. I also use a `context.should_abort_async` within the task to handle its cancellation. The issue appears with the following steps:

- a worker has started to process a task, marking its state as `doing`;
- in the middle of it, the task has been aborted from a different process, which changes the task state to `aborting`;
- after that, some expected error happens in the task itself (before the `should_abort_async` check), causing the worker to retry this task due to a retry strategy being used;
- during the retry `procrastinate_retry_job` SQL function will be used, which will raise  a `Job was not found or not in "doing" status` error;

This will shut down the worker process and will leave a task in the `procrastinate_jobs` table with an `aborting` status. If I just restart it, it won't also process newly added tasks for what it seems due to hanging an `aborting` task.

I understand that this might be a niche scenario, but I wonder if this behavior with the retry logic has been considered for a task abortion process. A quick and simple fix for this would be to mind the `aborting` status during job retry logic, but there is probably more to it.

Thanks for your attention!

### PR DESCRIPTION
**Pull request by medihack**
Closes #1131

### Successful PR Checklist:

- [x] Tests
  - [ ] (not applicable?)
- [ ] Documentation
  - [ ] (not applicable?)

#### PR label(s): 
  - [ ] https://github.com/procrastinate-org/procrastinate/labels/PR%20type%3A%20breaking%20%F0%9F%92%A5
  - [ ] https://github.com/procrastinate-org/procrastinate/labels/PR%20type%3A%20feature%20%E2%AD%90%EF%B8%8F
  - [x] https://github.com/procrastinate-org/procrastinate/labels/PR%20type%3A%20bugfix%20%F0%9F%95%B5%EF%B8%8F
  - [ ] https://github.com/procrastinate-org/procrastinate/labels/PR%20type%3A%20miscellaneous%20%F0%9F%91%BE
  - [ ] https://github.com/procrastinate-org/procrastinate/labels/PR%20type%3A%20dependencies%20%F0%9F%A4%96
  - [ ] https://github.com/procrastinate-org/procrastinate/labels/PR%20type%3A%20documentation%20%F0%9F%93%9A

This time, a job with the status `aborting` of a task that raises is set to `failed` (regardless of whether it should be retried or not). It feels a bit more consistent to me, but somehow, there is no optimal way. If we set a to-be-retried job in status `aborting` to `aborted` when the task raises, but one to `failed` that should not be retried, it seems somehow inconsistent to me.

@ewjoachim A note in the documentation is still missing. Just want to hear what you think.

### CODE REVIEW

File: tests/unit/test_worker.py [hunk 1]
```diff
@@ -385,6 +385,7 @@ def job_func(a, b):  # pylint: disable=unused-argument
         task_name="job",
         queue="yay",
     )
+    app.job_manager.get_job_status_async = mocker.AsyncMock(return_value="doing")
```
Comment (ewjoachim) [inline tests/unit/test_worker.py:388]: Just curious: what happens if we don't mock? The InMemory implementation doesn't answer correctly ?

**Review by ewjoachim — APPROVED** [review]
That's great! I believe there may be a bit of conflict when we merge the worker refactor, though. 

Thank you so much for the time you spent and doing it all over again after my comment!

File: tests/unit/test_worker.py (hunk 1 above)
Comment (medihack) [reply tests/unit/test_worker.py:388]: It'll throw. The job is not put into the worker queue. The `run_job` method is tested in isolation.

**Comment by medihack** [comment]
> That's great! I believe there may be a bit of conflict when we merge the worker refactor, though.

@onlyann When I commit this (later today), can you include it in the worker refactoring?

> Thank you so much for the time you spent and doing it all over again after my comment!

No prob, the comment was very relevant. And I am happy that we don't need a migration 🙂.

File: tests/unit/test_worker.py (hunk 1 above)
Comment (ewjoachim) [reply tests/unit/test_worker.py:388]: Ah, makes sense!

**Comment by onlyann** [comment]
That's not a problem, I will resolve the conflict.

I think it's fine to merge this fix but wanted to highlight that there can still be a race condition where the job is changed to "aborting" between the check and the moment the job is set to retry.

Something we can look at improving if we end up removing that extra status in v3

**Comment by medihack** [comment]
Alright, I will write a note in the documentation, merge it, and make a patch release.

**Comment by ewjoachim** [comment]
:+1:

**Pull request merged**